### PR TITLE
Balances cache optimization

### PR DESCRIPTION
**Pull request by polonzer**
https://lykkex.atlassian.net/browse/LWDEV-9147

- Changed cache serializatio into Protobuf.
- Changed cache structure for into HashSet.

### CODE REVIEW

File: src/Lykke.Service.Balances.Services/Wallet/CachedWalletModel.cs [hunk 1]
```diff
@@ -1,13 +1,18 @@
 ﻿using Lykke.Service.Balances.Core.Domain;
 using MessagePack;
+using ProtoBuf;
 
 namespace Lykke.Service.Balances.Services.Wallet
 {
+    [ProtoContract]
     [MessagePackObject(keyAsPropertyName: true)]
```
Comment (dmitrysmall) [inline src/Lykke.Service.Balances.Services/Wallet/CachedWalletModel.cs]: Do we still need this attribute?

File: src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs [hunk 2]
```diff
@@ -23,32 +25,59 @@ public class CachedWalletsRepository : ICachedWalletsRepository
...
+            [NotNull] ILogFactory logFactory,
+            [NotNull] IDatabase redisDatabase,
+            [NotNull] string partitionKey)
         {
-            _cache = cache ?? throw new ArgumentNullException(nameof(cache));
             _repository = repository ?? throw new ArgumentNullException(nameof(repository));
             _cacheExpiration = cacheExpiration;
+            _redisDatabase = redisDatabase ?? throw new ArgumentNullException(nameof(redisDatabase));
+            _partitionKey = partitionKey ?? throw new ArgumentNullException(nameof(partitionKey));
             _log = logFactory.CreateLog(this);
```
Comment (dmitrysmall) [inline src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs:37]: Shouldn't we check `logFactory` for `null` value like others?

File: src/Lykke.Service.Balances.Services/TotalBalancesService.cs [hunk 3]
```diff
@@ -9,17 +9,18 @@ namespace Lykke.Service.Balances.Services
...
     public class TotalBalancesService : ITotalBalancesService
     {
-        private static readonly string PartitionKey = "SpotClientBalances:total";
         private readonly IDatabase _redisDatabase;
+        private readonly string _partitionKey;
 
-        public TotalBalancesService(IDatabase redisDatabase)
+        public TotalBalancesService(IDatabase redisDatabase, string partitionKey)
         {
             _redisDatabase = redisDatabase;
```
Comment (dmitrysmall) [inline src/Lykke.Service.Balances.Services/TotalBalancesService.cs]: Do we need null-checks here like in `CachedWalletsRepository` ctor?

File: src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs [hunk 4]
```diff
@@ -23,32 +25,59 @@ public class CachedWalletsRepository : ICachedWalletsRepository
...
+            catch (RedisConnectionException ex)
+            {
+                _log.Warning("Redis cache is not available", ex);
+            }
+
+            var result = (await _repository.GetAsync(walletId)).Select(CachedWalletModel.Create).ToArray();
+
+            try
+            {
+                await _redisDatabase.HashSetAsync(
```
Comment (dmitrysmall) [inline src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs:60]: We have an `ILog` private instance here but do not pass it into Redis extension method. May be, we actually shoud, to have some more details about possible failure.

File: src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs [hunk 5]
```diff
@@ -58,9 +87,7 @@ public async Task UpdateBalanceAsync(string walletId, string assetId, decimal ba
             var updated = await _repository.UpdateBalanceAsync(walletId, wallet, updateSequenceNumber);
             if (updated)
             {
-                var key = GetAssetBalanceCacheKey(walletId, assetId);
-                await _cache.TrySetAsync(key, wallet, slidingExpiration: _cacheExpiration, log: _log);
-                await _cache.TryRemoveAsync(GetAllBalancesCacheKey(walletId));
+                await _redisDatabase.TryHashSetAsync(GetCacheKey(walletId), assetId, wallet);
```
Comment (dmitrysmall) [inline src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs]: The same about logger usage.

File: src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs (hunk 4 above)
Comment (polonzer) [reply src/Lykke.Service.Balances.Services/Wallet/CachedWalletsRepository.cs:60]: Here we use the Redis original method; it does not contain ILog parameter.

**Review by dmitrysmall — APPROVED** [review]

**Pull request merged**